openedx_tagging/core/tagging/api.py

```diff
@@ -100,5 +116,180 @@ def tag_object(
+                raise ValueError(
+                    _(
+                        f"Invalid JSON format: Missing 'tags' list."
+                    )
+                )
+            tags_data = tags_data.get('tags')
+        else:
+            raise ValueError(
+                _(
+                    f"Invalid format: {format}"
+                )
+            )
+    except ValueError as e:
+        raise e
+    finally:
+        tags.close()
+
+
+    new_tags = []
+    updated_tags = []
+
+    def create_update_tag(tag):
+        """
+        Function to create a new Tag or update an existing one.
+
+        This function keeps a creation/update history with `new_tags` and `updated_tags`,
+        a same tag can't be created/updated in a same taxonomy import.
+        Also, recursively, creates the parents of the `tag`.
+
+        Returns the created/updated Tag.
+        Raise KeyError if 'id' or 'name' don't exist on `tag`
+        """
+
+        tag_id = tag['id']
+        tag_name = tag['name']
+        tag_parent_id = tag.get('parent_id')
+        tag_parent_name = tag.get('parent_name')
+
+        # Check if the tag has not already been created or updated
+        if tag_id not in new_tags and tag_id not in updated_tags:
+            try:
+                # Update tag
+                tag_instance = taxonomy.tag_set.get(external_id=tag_id)
+                tag_instance.value = tag_name
+
+                if tag_instance.parent and (not tag_parent_id or not tag_parent_name):
+                    # if there is no parent in the data import
+                    tag_instance.parent = None
+                updated_tags.append(tag_id)
+            except Tag.DoesNotExist:
+                # Create tag
+                tag_instance = Tag(
+                    taxonomy=taxonomy,
+                    value=tag_name,
+                    external_id=tag_id,    
+                )
+                new_tags.append(tag_id)
+
+            if tag_parent_id and tag_parent_name:
+                # Parent creation/update
+                parent = create_update_tag({'id': tag_parent_id, 'name': tag_parent_name})
+                tag_instance.parent = parent
+
+            tag_instance.save()
+            return tag_instance
+        else:
+            # Returns the created/updated tag from history
+            return taxonomy.tag_set.get(external_id=tag_id)
+
+    # Create and update tags
+    with transaction.atomic():
+        # Delete all old Tags linked to the taxonomy
+        if replace:
+            Tag.objects.filter(taxonomy=taxonomy).delete()
+        
+        for tag in tags_data:
+            try:
+                create_update_tag(tag)
+            except KeyError as e:
+                key = e.args[0]
+                raise ValueError(
+                    _(
+                        f"Invalid JSON format: Missing '{key}' on a tag ({tag})"
+                    )
+                )
+        resync_tags()
```

Comment:
I [added an optional parameter to resync_tags](https://github.com/openedx/openedx-learning/pull/57#discussion_r1237422584) which will speed up this call:

```suggestion
        resync_tags(ObjectTag.objects.filter(taxonomy=taxonomy))
```

You'll need to merge the latest changes from https://github.com/openedx/openedx-learning/pull/57 over here to use it though.